crates/bevy_math/src/primitives/dim2.rs

```diff
@@ -0,0 +1,168 @@
+    /// Create a direction from a [Vec2] that is already normalized
+    pub fn from_normalized(value: Vec2) -> Self {
+        debug_assert!(value.is_normalized());
+        Self(value)
+    }
+}
+
+impl std::ops::Deref for Direction2d {
+    type Target = Vec2;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// A circle primitive
+#[derive(Clone, Copy, Debug)]
+pub struct Circle {
+    /// The radius of the circle
+    pub radius: f32,
+}
+impl Primitive2d for Circle {}
+
+/// An unbounded plane in 2D space. It forms a separating surface trough the origin,
+/// stretching infinitely far
+#[derive(Clone, Copy, Debug)]
+pub struct Plane2d {
+    /// The normal of the plane, the plane will be placed perpendicular to this direction
+    pub normal: Direction2d,
+}
+impl Primitive2d for Plane2d {}
+
+/// An infinite line along a direction in 2D space.
+///
+/// For a finite line: [`LineSegment2d`]
+#[derive(Clone, Copy, Debug)]
+pub struct Line2d {
+    /// The direction of the line, the line extends infinitely in both the positive
+    /// and negative of this direction
+    pub direction: Direction2d,
+}
+impl Primitive2d for Line2d {}
+
+/// A section of a line along a direction in 2D space.
+#[derive(Clone, Debug)]
+pub struct LineSegment2d {
+    /// The direction of the line
+    pub direction: Direction2d,
+    /// Half the length of the line segment, the segment extends by this amount in both
+    /// the in both the positive and negative direction
+    pub half_length: f32,
+}
+impl Primitive2d for LineSegment2d {}
+
+impl LineSegment2d {
+    /// Get a line segment and translation from a start and end position of a line segment
+    pub fn from_start_end(start: Vec2, end: Vec2) -> (Self, Vec2) {
+        let diff = end - start;
+        let length = diff.length();
+        (
+            Self {
+                direction: Direction2d::from_normalized(diff / length),
+                half_length: length / 2.,
+            },
+            (start + end / 2.),
+        )
+    }
+
+    /// Get the start position of the line
+    pub fn get_start_pos(&self) -> Vec2 {
+        *self.direction * -self.half_length
+    }
+
+    /// Get the end position of the line
+    pub fn get_end_pos(&self) -> Vec2 {
+        *self.direction * self.half_length
+    }
+}
+
+/// A series of connected line segments in 2D space.
+///
+/// For a version without generics: [`BoxedPolyline2d`]
+#[derive(Clone, Debug)]
+pub struct Polyline2d<const N: usize> {
+    /// The vertices of the polyline
+    pub vertices: [Vec2; N],
+}
+impl<const N: usize> Primitive2d for Polyline2d<N> {}
+
+/// A series of connected line segments in 2D space.
+///
+/// For a version without alloc: [`Polyline2d`]
+#[derive(Clone, Debug)]
+pub struct BoxedPolyline2d {
+    /// The vertices of the polyline
+    pub vertices: Box<[Vec2]>,
+}
+impl Primitive2d for BoxedPolyline2d {}
+
+/// A triangle primitive
+#[derive(Clone, Debug)]
+pub struct Triangle {
```

Comment:
I didn't give it a `2d` suffixed because there is currently no `Triangle3d`, but it might be reasonable to add one?

Reply:
Yes I think this is a reasonably common primitive to add one. Or at least assume it will be added sooner or later.

Reply:
I think having `Triangle3d` and `Quad3d` would make sense since they're defined by points which inherently depend on the dimension

Reply:
Not necessarily because they're flat 2D objects, and we define them relative to an oriented frame of reference. It depends if we include that orientation in the description of the primitive or not.

Reply:
I'll name it Triangle2d for now, and leave a 3d version for later if necessary. A Quad3d would be a bit more complex, since the representation can heavily affect how the primitive behaves, 4 vertices (like in the RFC) and a `Plane3d` + `Rectangle` would create very different behaviors.